nvflare/fuel/f3/communicator.py

```diff
@@ -48,15 +61,25 @@ def start(self):
         """
         self.conn_manager.start()
         log.debug(f"Communicator for local endpoint: {self.local_endpoint.name} is started")
+        _running_instances.add(self)
 
     def stop(self):
         """Stop the communicator and shutdown all the connections
 
         Raises:
             CommError: If any error encountered while shutting down
         """
+        if self.stopped:
+            return
+
         self.conn_manager.stop()
-        log.debug(f"Communicator for local endpoint: {self.local_endpoint.name} has stopped")
+        self.stopped = True
```

Comment:
should this line be before self.conn_manager.stop()?

Reply:
If stop() throws exception, this will allow exit_func to try it again.